### PR TITLE
Sends and labels `V*ReplicationClass`, `V*SnapClass`, `StorageClass` to client

### PR DESCRIPTION
**Pull request by raaizik**
## Changes

- Sends two VRCs (one for image flattening) and one VGRC. Depends on #2620 
- Labels V*RepClass, V*SnapClass and StorageClass

[RHSTOR-5753](https://issues.redhat.com/browse/RHSTOR-5753), [RHSTOR-5794](https://issues.redhat.com/browse/RHSTOR-5794), [RHSTOR-5754](https://issues.redhat.com/browse/RHSTOR-5754)

### CODE REVIEW

**Review by rewantsoni — CHANGES_REQUESTED** [review]
For Volume\*ReplicationClass we are sending only the parameters field like we did for Volume\*SnapshotClass and StorageClass, Should we send the entire spec for it or maybe the entire object that will allow us to add labels/annotations required (ramen labels, reclaimspace annotation) from the provider itself and the client would be responsible for adding missing fields like ProvisionerName and namespace for secrets and then creating/upating it.

@nb-ohad WDYT?

File: services/provider/server/server.go
Comment (Rakshith-R) [inline services/provider/server/server.go:696]:
```suggestion
					Name: "ceph-rbd-flatten",
```

Please make sure everything matches with https://github.com/red-hat-storage/odf-multicluster-orchestrator/blob/b3df88c23b06d80515f27a36f656296e2d413681/controllers/drpolicy_controller.go#L212

Comment (raaizik) [reply services/provider/server/server.go:696]: I notice the name's suffix is a hashed scheduling interval. Since I am watching the new DRClusterConfig resource (that provides the sched intervals) from the client op, I'd need to override this field there. But I guess it doesn't matter on your end because all you care about is that the client op sends the correct name.

**Review by Rakshith-R — COMMENTED** [review]
`"replication.storage.openshift.io/is-default-class": "true",`
This should be added as a annotation 

and 
`"replication.storage.openshift.io/flatten-mode": "force",`

This as a label

hope both of the above will be handled accordingly in consumer side.

Other than the above, everything else looks good to me.

**Comment by raaizik** [comment]
> `"replication.storage.openshift.io/is-default-class": "true",` This should be added as a annotation
> 
> and `"replication.storage.openshift.io/flatten-mode": "force",`
> 
> This as a label
> 
> hope both of the above will be handled accordingly in consumer side.
> 
> Other than the above, everything else looks good to me.

I'm aware and it will.

**Comment by nb-ohad** [comment]
> For Volume*ReplicationClass we are sending only the parameters field like we did for Volume*SnapshotClass and StorageClass, Should we send the entire spec for it or maybe the entire object that will allow us to add labels/annotations required (ramen labels, reclaimspace annotation) from the provider itself and the client would be responsible for adding missing fields like ProvisionerName and namespace for secrets and then creating/upating it.
> 
> @nb-ohad WDYT?

I will start with the later proposal first:
> maybe the entire object that will allow us to add labels/annotations required 

We do not want to do that, an object contains not just desired state but also status and some metadata that is local to the provider cluster. In internal discussions, we already acknowledged the need to send labels and annotations but not as part of the serialized spec we send (the data field of the ExternalResource struct). The plan is to add it to the metadata of the `ExternalResource` object, similar to the kind and name. Until that is implemented we will keep on having logic in the client to add the missing metadata. 

> Should we send the entire spec
This depends on the structure of the type, we originally used `data` because that was the desired state for configmaps and secrets. If the `Volume*ReplicationClass` has a formal spec field it will make sense to serialize it instead of a `map[string]string`. 
As a matter of fact, we are already doing something similar for the `ClusterResourceQuota` CR (https://github.com/red-hat-storage/ocs-operator/blob/main/services/provider/server/server.go#L376-L399)

**Comment by rewantsoni** [comment]
/hold
Holding as per https://github.com/red-hat-storage/ocs-operator/pull/2620#discussion_r1749185716
As the comment requests to add metadata (labels and annotation) and we are sending the labels and annotations both in the Data(Spec) field.

**Comment by raaizik** [comment]
> /hold Holding as per [#2620 (comment)](https://github.com/red-hat-storage/ocs-operator/pull/2620#discussion_r1749185716) As the comment requests to add metadata (labels and annotation) and we are sending the labels and annotations both in the Data(Spec) field.

Can unhold now 👍

**Review by rewantsoni — CHANGES_REQUESTED** [review]
Should we consider sending the spec of Volume*ReplicationClass instead of spec.parameter from the provider server?

**Comment by raaizik** [comment]
> Should we consider sending the spec of Volume*ReplicationClass instead of spec.parameter from the provider server?

I think we've had a previous discussion regarding this or similar. IMO the server should send only the info that's required and not the entire spec.

File: services/provider/server/server_test.go
Comment (rewantsoni) [inline services/provider/server/server_test.go:1142]:
```suggestion
		assert.True(t, reflect.DeepEqual(extResource.Labels, mockResoruce.Labels))
		assert.True(t, reflect.DeepEqual(extResource.Annotations, mockResoruce.Annotations))
```

File: services/provider/server/server_test.go
Comment (rewantsoni) [inline services/provider/server/server_test.go:359]: We could use assert.Equal() it uses reflect.DeepEqual internally

File: services/provider/server/server.go
Comment (rewantsoni) [inline services/provider/server/server.go:799]: from what I recall, we don't require this annotation for flatten VRC

File: services/provider/server/server.go
Comment (rewantsoni) [inline services/provider/server/server.go:824]: from what I recall, we don't require this annotation for flatten VGRC

File: services/provider/server/server.go
Comment (rewantsoni) [inline services/provider/server/server.go:914]: We don't need to differentiate between the Replication and Mirroring flag, we can have one mirroring flag (which will check the cephblockpool) and add labels based on that.

File: services/provider/server/server.go
Comment (rewantsoni) [inline services/provider/server/server.go:690]: StorageID is used to match V*C to StorageClass, we don't need it to be storageClaim UID, we can have it as storageRequest UID as well.

File: services/provider/server/server_test.go
Comment (raaizik) [reply services/provider/server/server_test.go:359]: Both `labels` and `annotations` are of type map, which means that in order to use `assert.Equal()` I'd need to convert from map to []byte which is not straightforward. Therefore I've chosen this assertion instead so it's compact as a one-liner.

File: services/provider/server/server.go
Comment (raaizik) [reply services/provider/server/server.go:799]: I'm pretty sure it is needed. @Rakshith-R

File: services/provider/server/server.go
Comment (raaizik) [reply services/provider/server/server.go:824]: See [reply](https://github.com/red-hat-storage/ocs-operator/pull/2727#discussion_r1771283490)

File: services/provider/server/server.go
Comment (raaizik) [reply services/provider/server/server.go:690]: We've already agreed that it'll be `string(r.storageClaim.UID)` in [#168](https://github.com/red-hat-storage/ocs-client-operator/pull/168). Eventually the value that will be assigned to this ID is the one from the client, so it actually doesn't matter what value is chosen on this end.

File: services/provider/server/server.go
Comment (raaizik) [reply services/provider/server/server.go:914]: From my perspective, there's no other way to handle these flags. Replication is `Vol*RepClass` specific, while Mirroring is aimed for both `Vol*RepClass` and `Vol*SnapClass`.

File: services/provider/server/server.go
Comment (rewantsoni) [reply services/provider/server/server.go:690]: As the concept for storageClaim might go away in the future because of convergence, we should pivot and not use it

File: services/provider/server/server.go
Comment (rewantsoni) [reply services/provider/server/server.go:799]: MCO is creating two VRC's one with flatten mode and one without flatten.
The one that is without flatten mode has the annotation the other one doesn't. See [here](https://github.com/red-hat-storage/odf-multicluster-orchestrator/blob/b3df88c23b06d80515f27a36f656296e2d413681/controllers/drpolicy_controller.go#L170-L195) and [here](https://github.com/red-hat-storage/odf-multicluster-orchestrator/blob/b3df88c23b06d80515f27a36f656296e2d413681/controllers/drpolicy_controller.go#L212-L226)
We are setting the annotations to be empty when creating the flatten mode VRC.

Also the same in mentioned in the design doc [Changes required](https://docs.google.com/document/d/1W9rnZNF82oUlIX7c0JJJrNDwx7pJfLU5X3suC-1jfqs/edit#heading=h.kf5l0za50hwu)

File: services/provider/server/server.go
Comment (rewantsoni) [reply services/provider/server/server.go:824]: See [reply](https://github.com/red-hat-storage/ocs-operator/pull/2727#discussion_r1771526698)

File: services/provider/server/server.go
Comment (rewantsoni) [reply services/provider/server/server.go:914]: Your replication flag depends on storageClusterPeer, whereas the mirroring flag is derived from CephBlockPool.

Enabling mirroring on the cephblockpool is the responsibility of the storageClusterPeer. Isn't both pointing to the same field eventually?

I think we should add the labels for rbd-based V*RC/V*SC and SC based only on the mirroring field. It is possible that a particular blockpool is not selected for mirroring from StorageClusterPeer CR. See [here](https://github.com/red-hat-storage/ocs-operator/blob/main/api/v1/storageclusterpeer_types.go#L50-L53)

I was going through the PR and I am not able to find the labels applied for CephFS-based V*SC and SC. Do we not need them?

File: services/provider/server/server.go
Comment (raaizik) [reply services/provider/server/server.go:690]: Even so, the final value assigned would be determined on the client.

Comment (rewantsoni) [reply services/provider/server/server.go:690]: Why do we need to determine the label value on the client if we can do it from the provider itself?

File: services/provider/server/server.go
Comment (rewantsoni) [reply services/provider/server/server.go:914]: >I am not aware of such thing. How does it fetch this flag exactly? Is it merged already?

This was discussed in the design discussion. Each StorageClusterPeer will enable mirroring on the cephblockpool based on the label selector.

For rbd we can use the mirroring field on the cephblockpool to add ramenRelated labels, we don't need another replication flag explicitly.

Comment (raaizik) [reply services/provider/server/server.go:914]: I think you're misunderstanding. I am using the Mirroring flag solely for the StorageID label. It's unrelated to RepID.

File: services/provider/server/server.go
Comment (raaizik) [reply services/provider/server/server.go:690]: We could, I'm saying with the current state of things this is what would happen. After this PR will get merged, the value assigned from here will be propagated to the client.

File: services/provider/server/server.go
Comment (rewantsoni) [reply services/provider/server/server.go:914]: Why do we need two flags for the same set of ramen labels? From what I understand we are creating a differentiation which we don't need. It will just complicate the things. 

As an example what if we have 2 cephblockpool a and b the storageclusterpeer is configured only for blockpool a.

There are two clients 1 and 2. 1 uses blockpool a, 2 uses blockpool b.

If we keep the flags separate,

For client 1 we will enable both replication and storage id label

For client 2 we will enable only replication id label

In the second case we don't require any ramen related labels to be added as that blockpool is not configured for mirroring. 

Hence we don't need to check for both storageclusterpeer and cephblockpool. Checking only the cephblockpool will work

File: services/provider/server/server.go
Comment (rewantsoni) [reply services/provider/server/server.go:690]: Yes, I agree that we could but I think as storageID is an arbitrary value used by ramen to cross-link the V*RC/V*SC and SC, the hash of storageRequest UID perfectly fits our use case. I don't think we need to change it further on the client. Let's wait for others to review this.

File: services/provider/server/server.go
Comment (raaizik) [reply services/provider/server/server.go:910]: > Each StorageClusterPeer will enable mirroring on the cephblockpool based on the label selector.

It doesn't make much sense to me to retrieve this flag by checking each SCP instead of a single CBP.

Comment (rewantsoni) [reply services/provider/server/server.go:910]: For rbd,
StorageClass will created from a radosnamesapce on a blockpool. From my perspective to populate the mirroring and replication flag, we should check if the mirroring is enabled for radosnamespace [spec.mirroring ](https://github.com/rook/rook/pull/14701)

For cephfs,
Since there is not need to perform any actions for this on ceph level unlike rbd, we should enable the mirroring and replication flag for this as soon as we have a storageClusterPeer present

File: services/provider/server/server.go
Comment (rewantsoni) [inline services/provider/server/server.go:713]: Could you change it to If rns.Spec.Mirroring is not nil then mirroring is enabled

File: services/provider/server/server.go
Comment (rewantsoni) [inline services/provider/server/server.go:886]: I think this is redundant

**Review by rewantsoni — CHANGES_REQUESTED** [review]

File: services/provider/server/server.go
Comment (raaizik) [reply services/provider/server/server.go:713]: I think this is more precise check (it checks both)

File: services/provider/server/server.go
Comment (raaizik) [reply services/provider/server/server.go:886]: This is the flag for setting up those IDs

File: services/provider/server/server.go
Comment (rewantsoni) [reply services/provider/server/server.go:766]: @raaizik I think this was missed

**Review by rewantsoni — CHANGES_REQUESTED** [review]

**Comment by rewantsoni** [comment]
LGTM

File: services/provider/server/server.go
Comment (rewantsoni) [inline services/provider/server/server.go:861]:
```suggestion
					Labels: getExternalResourceLabels("VolumeSnapshotClass", replicationEnabled, false, replicationID, storageID),
				},
```

File: services/provider/server/server.go
Comment (rewantsoni) [inline services/provider/server/server.go:868]:
```suggestion
					Labels: getExternalResourceLabels("VolumeGroupSnapshotClass", replicationEnabled, false, replicationID, storageID),
```

**Review by rewantsoni — CHANGES_REQUESTED** [review]

File: services/provider/server/server.go
Comment (raaizik) [reply services/provider/server/server.go]: The naming will be handled on consumer's end. Could you please resolve? @Rakshith-R

Comment (Rakshith-R) [reply services/provider/server/server.go]: I think i already mentioned it looks good to me here 
https://github.com/red-hat-storage/ocs-operator/pull/2727#pullrequestreview-2224881050

File: services/provider/server/server.go
Comment (Madhu-1) [inline services/provider/server/server.go:798]: we dont have voluemegroupreplicationclass in 4.18 isnt it? can we remove this and add it later?

File: services/provider/server/server.go
Comment (Madhu-1) [inline services/provider/server/server.go:868]: IMO we should use consts for all class kinds, we are passing it as string in multiple places.

File: services/provider/server/server.go
Comment (Madhu-1) [inline services/provider/server/server.go:902]: can we have a const for the label keys?

File: services/provider/server/server.go
Comment (raaizik) [reply services/provider/server/server.go:868]: If you mean replace it in all places not just the function call -- I did that at first, but @nb-ohad opposed to the idea

Comment (Madhu-1) [reply services/provider/server/server.go:849]: Yes we can easily make mistake when using string in multiple places. we need to have a var or const for it.

**Review by Madhu-1 — APPROVED** [review]
LGTM except https://github.com/red-hat-storage/ocs-operator/pull/2727/files#r1830834991

File: services/provider/server/server.go
Comment (raaizik) [reply services/provider/server/server.go]: @rewantsoni

**Pull request merged**
